File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/LibraryLoader.kt
package com.bockmedia.console.domain.model

import android.content.Context
import com.bockmedia.console.data.repository.BockMediaRepository
import com.bockmedia.console.domain.model.HomeArtworkCache
import com.bockmedia.console.local.OfflineDownloadManager
import com.bockmedia.console.local.OfflineDownloadStore
import com.bockmedia.console.local.toPlayTarget
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

enum class LibraryFilter(val label: String) {
    All("All"),
    Playlists("Playlists"),
    Artists("Artists"),
    Albums("Albums"),
    Downloaded("Downloads"),
}

enum class LibraryViewMode { List, Grid }

enum class LibrarySort { Recents, Name }

enum class LibraryItemKind { Playlist, Artist, Album, Downloaded }

data class LibraryItem(
    val id: String,
    val title: String,
    val subtitle: String,
    val kind: LibraryItemKind,
    val playTarget: PlayTarget,
    val artPath: String? = null,
    val playlistId: String? = null,
    val artistName: String? = null,
    val albumName: String? = null,
    val sortDate: Long = 0L,
    val unplayed: Boolean = false,
    val year: Int? = null,
    val avgStars: Double? = null,
)

object LibraryLoader {
    private const val BROWSE_PLAYLIST_LIMIT = 500
    /** Page size for Library Artists/Albums infinite scroll. */
    const val BROWSE_PAGE_SIZE = 60

    private fun albumSubtitle(album: com.bockmedia.console.data.api.dto.AlbumItem): String {
        val parts = mutableListOf<String>()
        album.year?.takeIf { it > 0 }?.let { parts.add(it.toString()) }
        album.artist?.takeIf { it.isNotBlank() }?.let { parts.add(it) }
        album.avgStars?.takeIf { it > 0 }?.let { stars ->
            parts.add("★ ${"%.1f".format(stars)}")
        }
        if (parts.isEmpty() && album.tracks > 0) parts.add("${album.tracks} tracks")
        return parts.joinToString(" · ")
    }

    suspend fun loadArtistPage(
        repository: BockMediaRepository,
        page: Int,
        search: String = "",
    ): Pair<List<LibraryItem>, Int> {
        val resp = runCatching {
            repository.artists(page = page, search = search, limit = BROWSE_PAGE_SIZE)
        }.getOrNull() ?: return emptyList<LibraryItem>() to 0
        return artistItems(repository, resp.items) to resp.total
    }

    suspend fun loadAlbumPage(
        repository: BockMediaRepository,
        page: Int,
        search: String = "",
    ): Pair<List<LibraryItem>, Int> {
        val resp = runCatching {
            repository.albums(page = page, search = search, limit = BROWSE_PAGE_SIZE)
        }.getOrNull() ?: return emptyList<LibraryItem>() to 0
        return albumItems(repository, resp.items) to resp.total
    }

    private suspend fun artistItems(
        repository: BockMediaRepository,
        artists: List<com.bockmedia.console.data.api.dto.ArtistItem>,
    ): List<LibraryItem> = artists.map { artist ->
        artist.artPath?.let { repository.cacheArtistArtPath(artist.name, it) }
        LibraryItem(
            id = "ar-${artist.name}",
            title = artist.name,
            subtitle = "${artist.albums} albums · ${artist.tracks} songs",
            kind = LibraryItemKind.Artist,
            playTarget = PlayTarget.Artist(artist.name),
            artistName = artist.name,
            artPath = artist.artPath,
        )
    }

    private suspend fun albumItems(
        repository: BockMediaRepository,
        albums: List<com.bockmedia.console.data.api.dto.AlbumItem>,
    ): List<LibraryItem> = albums.map { album ->
        album.artPath?.let { repository.cacheArtPath(album.name, album.artist, it) }
        LibraryItem(
            id = "al-${album.name}\u0000${album.artist.orEmpty()}",
            title = album.name,
            subtitle = albumSubtitle(album),
            kind = LibraryItemKind.Album,
            playTarget = PlayTarget.Album(album.name, album.artist),
            albumName = album.name,
            artistName = album.artist,
            artPath = album.artPath,
            unplayed = album.unplayed,
            year = album.year,
            avgStars = album.avgStars,
        )
    }

    /** Playlists + offline; artists/albums paginate on demand in LibraryScreen. */
    suspend fun loadBuckets(
        repository: BockMediaRepository,
        context: Context,
    ): LibraryData = coroutineScope {
        val playlistsDef = async {
            runCatching {
                repository.playlists(search = "", limit = BROWSE_PLAYLIST_LIMIT, memberScoped = true).items
            }.getOrDefault(emptyList()).also { items ->
                items.forEach { pl ->
                    pl.artPath?.takeIf { it.isNotBlank() }?.let { HomeArtworkCache.storePlaylistPath(pl.id, it) }
                }
            }.map { pl ->
                LibraryItem(
                    id = "pl-${pl.id}",
                    title = pl.name,
                    subtitle = "${pl.tracks ?: 0} songs",
                    kind = LibraryItemKind.Playlist,
                    artPath = pl.artPath,
                    playTarget = PlayTarget.Playlist(pl.id, pl.name),
                    playlistId = pl.id,
                    sortDate = 0L,
                )
            }
        }
        val offlineDef = async {
            OfflineDownloadManager.refresh(context)
            val store = OfflineDownloadStore(context)
            store.listManifests()
                .filter { manifest -> store.isCollectionComplete(manifest) }
                .map { manifest ->
                    LibraryItem(
                        id = "dl-${manifest.id}",
                        title = manifest.title,
                        subtitle = "${manifest.tracks.size} tracks · Offline",
                        kind = LibraryItemKind.Downloaded,
                        playTarget = manifest.toPlayTarget(),
                        artPath = manifest.coverArtPath,
                        playlistId = manifest.sourcePlaylistId ?: manifest.legacyPlaylistId,
                        sortDate = manifest.lastSyncedAtMs.takeIf { it > 0 } ?: manifest.downloadedAtMs,
                    )
                }
        }

        LibraryData(
            playlists = playlistsDef.await(),
            artists = emptyList(),
            albums = emptyList(),
            offline = offlineDef.await(),
        )
    }

    /** Search hits one category at a time (server-side filter). */
    suspend fun search(
        repository: BockMediaRepository,
        context: Context,
        filter: LibraryFilter,
        query: String,
    ): List<LibraryItem> = coroutineScope {
        val q = query.trim()
        if (q.isEmpty()) return@coroutineScope emptyList()
        when (filter) {
            LibraryFilter.Playlists, LibraryFilter.All -> {
                val items = runCatching {
                    repository.playlists(search = q, limit = BROWSE_PLAYLIST_LIMIT, memberScoped = true).items
                }.getOrDefault(emptyList())
                runCatching { repository.prefetchPlaylistCoverPaths(items.map { it.id }) }
                items.map { pl ->
                    LibraryItem(
                        id = "pl-${pl.id}",
                        title = pl.name,
                        subtitle = "${pl.tracks ?: 0} songs",
                        kind = LibraryItemKind.Playlist,
                        artPath = pl.artPath,
                        playTarget = PlayTarget.Playlist(pl.id, pl.name),
                        playlistId = pl.id,
                    )
                }
            }
            LibraryFilter.Artists -> loadArtistPage(repository, page = 1, search = q).first
            LibraryFilter.Albums -> loadAlbumPage(repository, page = 1, search = q).first
            LibraryFilter.Downloaded -> {
                OfflineDownloadManager.refresh(context)
                OfflineDownloadStore(context).listManifests()
                    .filter { it.title.contains(q, ignoreCase = true) }
                    .map { manifest ->
                        LibraryItem(
                            id = "dl-${manifest.id}",
                            title = manifest.title,
                            subtitle = "${manifest.tracks.size} tracks · Offline",
                            kind = LibraryItemKind.Downloaded,
                            playTarget = manifest.toPlayTarget(),
                            artPath = manifest.coverArtPath,
                            playlistId = manifest.sourcePlaylistId ?: manifest.legacyPlaylistId,
                        )
                    }
            }
        }
    }
}

File: ios/BockMedia/Domain/LibraryLoader.swift
import Foundation

enum LibraryFilter: String, CaseIterable, Identifiable {
    case all = "All"
    case playlists = "Playlists"
    case artists = "Artists"
    case albums = "Albums"
    case tracks = "Tracks"
    case downloaded = "Downloaded"

    var id: String { rawValue }
    var label: String { rawValue }
}

enum LibraryViewMode { case list, grid }

enum LibrarySort { case recents, name }

enum LibraryItemKind { case playlist, artist, album, track, downloaded }

struct LibraryItem: Identifiable, Hashable {
    let id: String
    let title: String
    let subtitle: String
    let kind: LibraryItemKind
    let playTarget: PlayTarget
    let artPath: String?
    let playlistId: String?
    let artistName: String?
    let albumName: String?
    let sortDate: Int64
    let unplayed: Bool

    init(
        id: String,
        title: String,
        subtitle: String,
        kind: LibraryItemKind,
        playTarget: PlayTarget,
        artPath: String? = nil,
        playlistId: String? = nil,
        artistName: String? = nil,
        albumName: String? = nil,
        sortDate: Int64 = 0,
        unplayed: Bool = false
    ) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.kind = kind
        self.playTarget = playTarget
        self.artPath = artPath
        self.playlistId = playlistId
        self.artistName = artistName
        self.albumName = albumName
        self.sortDate = sortDate
        self.unplayed = unplayed
    }
}

struct LibraryData {
    let playlists: [LibraryItem]
    let artists: [LibraryItem]
    let albums: [LibraryItem]
    let offline: [LibraryItem]

    func forFilter(_ filter: LibraryFilter) -> [LibraryItem] {
        switch filter {
        case .all: return playlists + artists + albums + offline
        case .playlists: return playlists
        case .artists: return artists
        case .albums: return albums
        case .tracks: return []
        case .downloaded: return offline
        }
    }
}

enum LibraryLoader {
    private static let browsePlaylistLimit = 80
    private static let browseArtistLimit = 60
    private static let browseAlbumLimit = 60
    private static let searchLimit = 100

    static func loadBuckets(repository: BockMediaRepository) async -> LibraryData {
        async let playlistsTask = loadPlaylists(repository: repository, search: "", limit: browsePlaylistLimit)
        async let artistsTask = loadArtists(repository: repository, search: "", limit: browseArtistLimit)
        async let albumsTask = loadAlbums(repository: repository, search: "", limit: browseAlbumLimit)
        async let offlineTask = loadOffline()

        return LibraryData(
            playlists: await playlistsTask,
            artists: await artistsTask,
            albums: await albumsTask,
            offline: await offlineTask
        )
    }

    static func unifiedSearch(repository: BockMediaRepository, query: String, source: String? = nil) async throws -> SearchResponse {
        try await UnifiedSearchCoordinator.fetch(repository: repository, query: query, source: source)
    }

    static func search(
        repository: BockMediaRepository,
        filter: LibraryFilter,
        query: String
    ) async -> [LibraryItem] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return [] }
        switch filter {
        case .playlists:
            return await loadPlaylists(repository: repository, search: q, limit: searchLimit)
        case .all:
            return []
        case .artists:
            return await loadArtists(repository: repository, search: q, limit: searchLimit, page: 1)
        case .albums:
            return await loadAlbums(repository: repository, search: q, limit: searchLimit, page: 1)
        case .tracks:
            return await loadTracks(repository: repository, search: q, limit: searchLimit, page: 1)
        case .downloaded:
            await OfflineDownloadManager.shared.refresh()
            return OfflineDownloadStore().listManifests()
                .filter { $0.title.localizedCaseInsensitiveContains(q) }
                .map(offlineItem(manifest:))
        }
    }

    static func loadPage(
        repository: BockMediaRepository,
        filter: LibraryFilter,
        search: String,
        page: Int,
        limit: Int = 60
    ) async -> (items: [LibraryItem], total: Int) {
        switch filter {
        case .artists:
            let resp = try? await repository.artists(page: page, limit: limit, search: search)
            let items = resp?.items.map { artist in
                LibraryItem(
                    id: "ar-\(artist.name.hashValue)",
                    title: artist.name,
                    subtitle: "\(artist.albums) albums · \(artist.tracks) songs",
                    kind: .artist,
                    playTarget: .artist(name: artist.name),
                    artPath: artist.artPath,
                    artistName: artist.name
                )
            } ?? []
            return (items, resp?.total ?? items.count)
        case .albums:
            let resp = try? await repository.albums(page: page, limit: limit, search: search)
            let items = resp?.items.map { album in
                LibraryItem(
                    id: "al-\(album.name.hashValue)-\((album.artist ?? "").hashValue)",
                    title: album.name,
                    subtitle: [album.artist, album.year.map(String.init)].compactMap { $0 }.joined(separator: " · "),
                    kind: .album,
                    playTarget: .album(name: album.name, artist: album.artist),
                    artPath: album.artPath,
                    artistName: album.artist,
                    albumName: album.name,
                    unplayed: album.unplayed
                )
            } ?? []
            return (items, resp?.total ?? items.count)
        case .tracks:
            let resp = try? await repository.songs(page: page, limit: limit, search: search)
            let items = await loadTracks(repository: repository, search: search, limit: limit, page: page)
            return (items, resp?.total ?? items.count)
        default:
            return ([], 0)
        }
    }

    private static func loadPlaylists(repository: BockMediaRepository, search: String, limit: Int) async -> [LibraryItem] {
        let items = (try? await repository.playlists(search: search, limit: limit, memberScoped: true))?.items ?? []
        await repository.prefetchPlaylistCoverPaths(ids: items.map(\.id))
        return items.map { pl in
            LibraryItem(
                id: "pl-\(pl.id)",
                title: pl.name,
                subtitle: "\(pl.tracks) songs",
                kind: .playlist,
                playTarget: .playlist(id: pl.id, name: pl.name),
                artPath: pl.artPath,
                playlistId: pl.id
            )
        }
    }

    private static func loadArtists(repository: BockMediaRepository, search: String, limit: Int, page: Int = 1) async -> [LibraryItem] {
        (try? await repository.artists(page: page, limit: limit, search: search))?.items.map { artist in
            LibraryItem(
                id: "ar-\(artist.name.hashValue)",
                title: artist.name,
                subtitle: "\(artist.albums) albums · \(artist.tracks) songs",
                kind: .artist,
                playTarget: .artist(name: artist.name),
                artPath: artist.artPath,
                artistName: artist.name
            )
        } ?? []
    }

    private static func loadAlbums(repository: BockMediaRepository, search: String, limit: Int, page: Int = 1) async -> [LibraryItem] {
        (try? await repository.albums(page: page, limit: limit, search: search))?.items.map { album in
            LibraryItem(
                id: "al-\(album.name.hashValue)-\((album.artist ?? "").hashValue)",
                title: album.name,
                subtitle: [album.artist, album.year.map(String.init)].compactMap { $0 }.joined(separator: " · "),
                kind: .album,
                playTarget: .album(name: album.name, artist: album.artist),
                artPath: album.artPath,
                artistName: album.artist,
                albumName: album.name,
                unplayed: album.unplayed
            )
        } ?? []
    }

    static func loadTracks(repository: BockMediaRepository, search: String, limit: Int, page: Int = 1) async -> [LibraryItem] {
        (try? await repository.songs(page: page, limit: limit, search: search))?.items.compactMap { song in
            guard let path = song.path?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank,
                  let title = song.title?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank else {
                return nil
            }
            let subtitle = [song.artist, song.album].compactMap { $0?.nilIfBlank }.joined(separator: " · ")
            return LibraryItem(
                id: "tr-\(path)",
                title: title,
                subtitle: subtitle.isEmpty ? "Track" : subtitle,
                kind: .track,
                playTarget: .song(path: path, title: title),
                artPath: path,
                artistName: song.artist,
                albumName: song.album
            )
        } ?? []
    }

    private static func loadOffline() async -> [LibraryItem] {
        await OfflineDownloadManager.shared.refresh()
        return OfflineDownloadStore().listManifests()
            .filter { OfflineDownloadStore().isCollectionComplete($0) }
            .map(offlineItem(manifest:))
    }

    private static func offlineItem(manifest: OfflineCollectionManifest) -> LibraryItem {
        LibraryItem(
            id: "dl-\(manifest.id)",
            title: manifest.title,
            subtitle: "\(manifest.tracks.count) tracks · Offline",
            kind: .downloaded,
            playTarget: manifest.toPlayTarget(),
            artPath: manifest.coverArtPath,
            playlistId: manifest.sourcePlaylistId ?? manifest.legacyPlaylistId,
            sortDate: manifest.lastSyncedAtMs > 0 ? manifest.lastSyncedAtMs : manifest.downloadedAtMs
        )
    }
}

private extension String {
    var nilIfBlank: String? {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
}
